node/core/dispute-coordinator/src/initialized.rs

```diff
@@ -1186,6 +1191,89 @@ impl Initialized {
 
 		Ok(())
 	}
+
+	/// Firstly decrements spam slots for validators who voted on potential spam
+	/// candidates that are newly backed or included, and therefore no longer
+	/// potential spam.
+	/// 
+	/// Secondly queues participation if needed for freshly backed or included 
+	/// candidates. Backed candidates are added to best_effort queue while 
+	/// included are added to priority queue.
+	async fn handle_backed_included_events<Context>(
+		&mut self, 
+		ctx: &mut Context, 
+		overlay_db: &mut OverlayedBackend<'_, impl Backend>, 
+		now: u64,
+		candidate_events: Vec<CandidateEvent>,
+	)
+	{
+		let session = self.rolling_session_window.latest_session();
+		// Populate events contents, which allow us to queue participation without
+		// duplicated code
+		let mut events_contents: Vec<(CandidateReceipt, ParticipationPriority)> = Vec::new();
+		for event in candidate_events {
+			match event {
+				CandidateEvent::CandidateBacked(receipt, _, _, _) => {
+					events_contents.push((receipt, ParticipationPriority::BestEffort));
+				},
+				CandidateEvent::CandidateIncluded(receipt, _, _, _) => {
+					events_contents.push((receipt, ParticipationPriority::Priority));
+				}
+				_ => (),
+			}
+		}
+
+		for (receipt, priority) in events_contents {
+			// Clear spam slots
+			self.spam_slots.clear(&(session, receipt.hash()));
+
+			// Queue participation for freshly backed candidate. If dispute
```

Comment:
I think this code partially duplicates `process_on_chain_votes`. Importing backing votes is handled there.
What should happen here is only promoting best effort candidates to priority ones (in case they we included) and updating the spam slots.

Reply:
Yep, the logic that should always be the same between importing and this code should be factored out in sensible fashion ... so it will always stay consistent. E.g. define a function `fn should_participate(...)`.

Reply:
Good catch .... new backing votes are indeed already handled, due to chain scraping votes. :bulb:

Reply:
Or even better factor out the full participation logic into something like `fn participate_if_necessary(...)` with needed parameters. Then we can just call it on any currently disputed candidate after we updated scraped events.

Reply:
I still don't understand. I was with the impression that we should only move disputes from best effort to priority if they got included.
Why we can't scan best effort queue, check if something there is viable for priority (got included, etc), move it to priority and let the current code do its job? `process_active_leaves_update` in scraper knows which candidates are freshly included. The function with the same name in the participation module triggers the actual participation.

What am I missing?

Reply:
You are right! Given that also spam slots are handled by the scraping of backing votes, that is really everything that is missing. Good thinking!

So indeed, no need to fetch votes, no need to know about current disputes. Just bump any participation request in best-effort to priority on new included events. :1st_place_medal: 

That turned out nice :-) I would love a comment on the on chain vote processing that it accomplishes those goals implicitly though:

Apart from having any backing votes imported, the processing of on-chain backing votes has also the following side   effects on already ongoing disputes:

- Spam slots will be cleared for disputers of any candidate that is now known to be backed.
- Participation for previously unknown candidates will be triggered

Reply:
I'm wondering what my take away should be from this whole conversation implementation wise.

Rather than triggering any new `queue_participation` calls, perhaps this?

 I create a new function in `Participation` that takes a set of candidate hashes that have been freshly included and checks the best effort queue to see if any matching disputes need to be bumped up to priority.

Reply:
Yes, it should be enough to check if there are candidates in best effort queue which were included, move them to priority and let the rest of the code do its job.
It should happen between the  two lines I shared above. `self.scraper.process_active_leaves_update` will return the freshly scraped included events (you already did this). `self.participation.process_active_leaves_update` will  trigger participation. So rearranging the lists between these two calls should be enough.